### PR TITLE
Adds Features per Fix tracking

### PR DESCRIPTION
**Pull request by Cyberboss**
Related thread: https://tgstation13.org/phpBB/viewtopic.php?f=5&t=10057

This sets up @tgstation-server to track the amount of 'issues positive' PRs against the amount of 'issue negative' PRs that a have been *merged* from a user. The system is based on the PR's tags. If the amount goes negative new 'issue negative' PRs will have a comment posted on them to nag the user to fix more shit. Maintainers can always feel free to ignore/delete the comment or close the PR. All users start with a positive balance of whatever the configured rate is. This system also allows for neutral PRs which don't affect the balance.

@tgstation/commit-access 

# https://tools.tgstation13.org/pr_balances.json

### CODE REVIEW

File: tools/github_webhook_processor.php [hunk 1]
```diff
@@ -23,6 +23,9 @@
 //This is required as otherwise somebody could trick the script into leaking the api key.
 $hookSecret = '08ajh0qj93209qj90jfq932j32r';
 
+$prBalanceJson = '';	//Set this to the path you'd like the writable pr balance file to be stored, not setting it writes it to the working directory
+$featuresPerFix = 1;	//Number of allowed 'feature' PRs per merged 'fix' PR
```
Comment (Cyberboss) [inline tools/github_webhook_processor.php]: Discuss

File: tools/github_webhook_processor.php [hunk 2]
```diff
@@ -195,6 +198,12 @@ function handle_pr($payload) {
 	switch ($payload["action"]) {
 		case 'opened':
 			tag_pr($payload, true);
+			if(get_pr_code_friendliness($payload) === -1){
+				$balances = pr_balances();
+				$author = $payload['pull_request']['user']['login'];
+				if(isset($balances[$author]) && $balances[$author] < 0)
+					create_comment($payload, "You currently have a higher Feature/Fix pull request ratio than the configured rate of " . $featuresPerFix . ". Maintainers may close this PR at will. Fixing issues or improving the codebase will improve this score.");
```
Comment (Cyberboss) [inline tools/github_webhook_processor.php]: Discuss

File: tools/github_webhook_processor.php [hunk 3]
```diff
@@ -223,7 +233,89 @@ function handle_pr($payload) {
...
+		return array();
+}
+
+//returns 1 if it's a fix/feature/etc, 0 if it's a neutral, -1 if it's a Feature/Tweak/Balance
+function get_pr_code_friendliness($payload){
+	$labels = get_pr_labels_array($payload);
+	//doing one of these increases your positive score
+	$improvement_labels = array('Fix', 'Refactor', 'Code Improvement', 'Priority: High', 'Priority: CRITICAL', 'Atmospherics', 'Grammar and Formatting', 'Logging', 'Performance');
+	//doing one of these prevents your negative score from increasing
+	$neutral_labels = array('Tools', 'Map Edit', 'SQL', 'Documentation', 'Repository', 'Sound', 'Revert/Removal', 'UI', 'Sprites', 'Sound');
```
Comment (Cyberboss) [inline tools/github_webhook_processor.php]: Discuss

File: tools/github_webhook_processor.php [hunk 4]
```diff
@@ -223,7 +233,89 @@ function handle_pr($payload) {
...
+	else if($friendliness === 1)
+		$friendliness *= $featuresPerFix;
+
+	$author = $payload['pull_request']['user']['login'];
+	$balances = pr_balances();
+	if(!isset($balances[$author]))
+		$balances[$author] = $featuresPerFix;
+	$balances[$author] += $friendliness;
+	if($balances[$author] < 0)
+		create_comment($payload, "Your Fix/Feature pull request ratio has just gone below the configured amout of " . $featuresPerFix . " features per fix. Maintainers may close future Feature/Tweak/Balance PRs. Fixing issues or improving the codebase will improve this score.");
```
Comment (Cyberboss) [inline tools/github_webhook_processor.php]: Discuss

File: tools/github_webhook_processor.php (hunk 3 above)
Comment (RemieRichards) [inline tools/github_webhook_processor.php]: make maintainers immune to bad prs pls.

File: tools/github_webhook_processor.php [hunk 5]
```diff
@@ -223,7 +235,91 @@ function handle_pr($payload) {
...
+		return array();
+}
+
+//returns 1 if it's a fix/feature/etc, 0 if it's a neutral, -1 if it's a Feature/Tweak/Balance
+function get_pr_code_friendliness($payload){
+	$labels = get_pr_labels_array($payload);
+	//doing one of these increases your positive score
+	$improvement_labels = array('Fix', 'Refactor', 'Code Improvement', 'Priority: High', 'Priority: CRITICAL', 'Atmospherics', 'Grammar and Formatting', 'Logging', 'Performance');
+	//doing one of these prevents your negative score from increasing
+	$neutral_labels = array('Tools', 'Map Edit', 'SQL', 'Documentation', 'Repository', 'Revert/Removal', 'UI', 'Sprites', 'Sound');
```
Comment (Cyberboss) [inline tools/github_webhook_processor.php]: Discuss

**Comment by optimumtact** [comment]
It's stupid because we're not people's mums and we have no right to nag them about doing things they obviously don't want to do.

**Comment by optimumtact** [comment]
actually on a second thought I'm not so opposed to this as long as we nail the wording

**Comment by pubby** [comment]
My concern is that this does nothing to stop buggy PRs, and if anything incentivizes sneaking in bugs on purpose so that you can have trivial fixes ready to gain points.

File: tools/github_webhook_processor.php (hunk 1 above)
Comment (Tacolizard) [inline tools/github_webhook_processor.php]: might be good in theory, but will probably just spam people. at least 2 would be better

File: tools/github_webhook_processor.php (hunk 2 above)
Comment (Tacolizard) [inline tools/github_webhook_processor.php]: uh "maintainers may close this PR at will" is kinda harsh and could be abused. this system should really be 70% nagging and 30% actual consequences.

**Comment by MrStonedOne** [comment]
>'issues positive' PRs against the amount of 'issue negative' PRs

Explain what this means.

**Comment by KorPhaeron** [comment]
Biggest issue with this I can see is gaming the system with grammar fix PRs

Fixing a server crash will be worth the same as adding a comma 

Big refactors wont be worth any points either despite being code improvement

Fighting over competing fixes might be become a thing

**Comment by RemieRichards** [comment]
@KorPhaeron we can add scaling points and make grammar/spelling etc. neutral (there's no reason to do it already, and yet people still do)

and things like 'PRIORITY:HIGH' could be worth 5, instead of 1, it's all solvable, we just need to assign values to tags.

@Cyberboss some kind of config where tags are assigned like so:
"fix": 1 (Positive)
"grammar": 0 (Neutral)
"feature": -1 (Negative)
"high priority": 5
...
would be nice, then a PR's score is:
Highest Positive Tag (eg: high priority + fix is '5', not '1' or '6', highest and not sum so 'easy fix' + 'fix' isn't suddenly worth 2 points) - All Negative Tags ( + Neutral tags, technically, but they'd all be 0 so it'd moot)
eg: a high priority fix that included a feature (we've had dummer prs) would be:
5 - 1= 4

@tgstation/commit-access thoughts?

**Comment by pubby** [comment]
call the system CodiMarket

**Comment by RemieRichards** [comment]
no it's called the 
# PR-etty Good PR System.

File: tools/github_webhook_processor.php (hunk 5 above)
Comment (Cyberboss) [inline tools/github_webhook_processor.php]: @MrStonedOne in regards to your comment below, $improvement_labels gives you positive score, $neutral_labels give you zero score, all other labels give you -1 score

**Comment by Qbopper** [comment]
would be better is you just called it the PR-etty Good System or something, having PR in it twice is redundant

**Comment by RemieRichards** [comment]
the redundancy is the joke.

**Comment by Qbopper** [comment]
bad joke

**Comment by Cyberboss** [comment]
@tgstation/commit-access Here's my proposition for Remie's point system

Fix += 2
Refactor += 2
Code Improvement += 1 
Priority: High += 4
Priority: CRITICAL += 5
Atmospherics += 4
Logging += 1
Performance += 3
All of these are 0: Grammar and Formatting, Tools, Map Edit, SQL, Documentation, Repository, Revert/Removal, UI, Sprites, Sound
Everything else is -1

Thoughts?

**Comment by XDTM** [comment]
Might want to track if you're fixing your own bugs, i should not be rewarded with a +3 because i fucked up and made viruses only affect one person and then fixed it

**Comment by RemieRichards** [comment]
@XDTM I figured that didn't need saying, but yes definitely.
@Cyberboss sounds good.

**Comment by kevinz000** [comment]
goofmarket on github lets go

File: tools/github_webhook_processor.php [hunk 6]
```diff
@@ -23,6 +23,11 @@
 //This is required as otherwise somebody could trick the script into leaking the api key.
 $hookSecret = '08ajh0qj93209qj90jfq932j32r';
 
+$prBalanceJson = '';	//Set this to the path you'd like the writable pr balance file to be stored, not setting it writes it to the working directory
+$featuresPerFix = 1;	//Number of allowed 'feature' PRs per merged 'fix' PR
+//TODO: configure to pull from an org team maybe?
+$maintainers = array('AnturK', 'ChangelingRain', 'Cheridan', 'Cyberboss', 'Jordie0608', 'lzimann', 'KorPhaeron', 'Razharas', 'RemieRichards', 'WJohn');
```
Comment (optimumtact) [inline tools/github_webhook_processor.php]: hurtful code

File: tools/github_webhook_processor.php (hunk 2 above)
Comment (optimumtact) [inline tools/github_webhook_processor.php]: We can close your prs at will anyway taco

**Comment by optimumtact** [comment]
This seems ready apart from missing a few maintainers

**Comment by Jordie0608** [comment]
Am no huge fan; I think we should be nannying the code, not contributors.

**Comment by optimumtact** [comment]
This is at worst a harmless warning that can be ignored, and at best somethign that actually encourages people to make fixes.

what is there to lose?

**Comment by Jordie0608** [comment]
I agree it's not going to hurt to have unless it's misused, I just don't see much benefit either.

**Comment by AnturK** [comment]
Not sure on the whole scheme since i doubt it will effect any change but i guess we can try.

**Comment by Cyberboss** [comment]
@optimumtact Who is it missing?

**Comment by Cyberboss** [comment]
I left out doris intentionally, unless his position has changed?

**Comment by RemieRichards** [comment]
it didn't.
his job is taken by the robot now.

**Comment by RemieRichards** [comment]
Though I know exactly what oranges is referring to :^)

**Comment by Cyberboss** [comment]
OH LOL I DIDN't EVEN NOTICE

File: tools/github_webhook_processor.php [hunk 7]
```diff
@@ -270,7 +283,108 @@ function handle_pr($payload) {
...
+//return the assoc array of login -> balance for prs
+function pr_balances(){
+	$path = pr_balance_json_path();
+	if(file_exists($path))
+		return json_decode(file_get_contents($path), true);
+	else
+		return array();
+}
+
+//returns 1 if it's a fix/feature/etc, 0 if it's a neutral, -1 if it's a Feature/Tweak/Balance
```
Comment (RemieRichards) [inline tools/github_webhook_processor.php]: not true anymore

File: tools/github_webhook_processor.php [hunk 8]
```diff
@@ -270,7 +283,108 @@ function handle_pr($payload) {
...
+function get_pr_code_friendliness($payload, $oldbalance){
+	global $startingPRBalance;
+	$labels = get_pr_labels_array($payload);
+	//anything not in this list defaults to -1
```
Comment (RemieRichards) [inline tools/github_webhook_processor.php]: should default to doing nothing, and have both positive (already done) and negative need to be explicitly set

File: tools/github_webhook_processor.php [hunk 9]
```diff
@@ -270,7 +283,108 @@ function handle_pr($payload) {
...
+	);
+
+	$affecting = -1;
+	$is_neutral = FALSE;
+	foreach($labels as $l){
+		if($l == 'PRB: Reset') {	//sets back to starting balance
+			$affecting = max($affecting, $startingPRBalance - $oldbalance);
+			break;
+		}
+		else if($l == 'PRB: No Update') {	//no effect on balance
```
Comment (RemieRichards) [inline tools/github_webhook_processor.php]: would also remove the need for this block.

File: tools/github_webhook_processor.php [hunk 10]
```diff
@@ -232,11 +237,18 @@ function tag_pr($payload, $opened) {
 }
 
 function handle_pr($payload) {
+	global $featuresPerFix;
 	$action = 'opened';
 	$validated = validate_user($payload);
 	switch ($payload["action"]) {
 		case 'opened':
 			tag_pr($payload, true);
+			if(get_pr_code_friendliness($payload) === -1){
```
Comment (RemieRichards) [inline tools/github_webhook_processor.php]: Not true anymore!!

**Comment by lzimann** [comment]
Never have any opinion on this but don't care either way on how this goes.

**Comment by Cyberboss** [comment]
This is tested and ready @MrStonedOne

**Comment by Cyberboss** [comment]
I take it back, didn't test teams yet ree

**Comment by Cyberboss** [comment]
Tested, working, fixes another bug we missed @MrStonedOne

File: tools/github_webhook_processor.php [hunk 11]
```diff
@@ -268,7 +281,114 @@ function handle_pr($payload) {
...
+	$author = $payload['pull_request']['user']['login'];
+	if(is_maintainer($payload, $author))	//immune
+		return;
+	$balances = pr_balances();
+	if(!isset($balances[$author]))
+		$balances[$author] = $startingPRBalance;
+	$friendliness = get_pr_code_friendliness($payload, $balances[$author]);
+	$balances[$author] += $friendliness;
+	if($balances[$author] < 0)
+		create_comment($payload, 'Your Fix/Feature pull request delta is currently below zero (' . $balances[$author] . '). Maintainers may close future Feature/Tweak/Balance PRs. Fixing issues or helping to improve the codebase will raise this score.');
```
Comment (MrStonedOne) [inline tools/github_webhook_processor.php]: Make this only show if the merged pr is also negative.

If you get really low, and start doing fix prs to work your way back up, this message showing on merge of positive prs would be lame and unhelpful.

Comment (MrStonedOne) [reply tools/github_webhook_processor.php]: On that note, maybe make another message that shows if a author's balance was below 0, but is now above it.

**Comment by optimumtact** [comment]
so by my count we have
@RemieRichards for
@Cyberboss for
@Jordie0608 against
@AnturK against
@optimumtact doesn't care either way (formerly against)
@lzimann doesn't care either way

@ChangelingRain @Cheridan @KorPhaeron can at least one of you weigh in to split the tie.

**Comment by RemieRichards** [comment]
@optimumtact antur is willing to give it a try, so I'd say he's not actually 'against', thus putting this PR in the 'lets do this' category

**Comment by Iamgoofball** [comment]
@optimumtact so we're just going to believe remie at face value now? she could of been lying

**Comment by AnturK** [comment]
I'm not sure what about that was ambiguous ? I'm fine with this merge.

**Comment by lzimann** [comment]
goof trying to find an excuse to not have this so he doesn't have to do fixes

**Comment by Iamgoofball** [comment]
Cesar and Deitus

**Pull request merged**